### PR TITLE
{devel}[foss/2019b] PyTorch v1.6.0 w/ Python 3.7.4

### PR DESCRIPTION
**Pull request by boegel**
(created using `eb --new-pr`)

### CODE REVIEW

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node3404.kirlia.os - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) Gold 6240 CPU @ 2.60GHz (cascadelake), Python 2.7.5
See https://gist.github.com/0176a6d53ec62de97a25cd2511130577 for a full test report.

**Comment by boegel** [comment]
@boegelbot please test @ generoso

**Comment by verdurin** [comment]
Test report by @verdurin
**FAILED**
Build succeeded for 4 out of 5 (1 easyconfigs in this PR)
nuc.lan - Linux Fedora 32, x86_64, Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz, Python 3.8.5
See https://gist.github.com/a8b02ab605d83a10925b321c2f3ee390 for a full test report.

**Comment by boegel** [comment]
Tests are failing with `AssertionError: False is not true : Expected zero exit code but got -6` in @verdurin's test report.

Does this ring a bell for you @Flamefire?

**Comment by verdurin** [comment]
Individual failures are like this:
```
FAIL
test_backward_without_context (__main__.TensorPipeAgentDistAutogradTestWithSpawn) ... terminate called after throwing an instance of 'terminate called after throwing an instance of 'std::runtime_errorstd::runtime_error'
'
  what():  In initFromLoop at ../third_party/tensorpipe/tensorpipe/transport/uv/listener.cc:126 "rv < 0: invalid argument"
  what():  In initFromLoop at ../third_party/tensorpipe/tensorpipe/transport/uv/listener.cc:126 "rv < 0: invalid argument"
terminate called after throwing an instance of 'std::runtime_error'
  what():  In initFromLoop at ../third_party/tensorpipe/tensorpipe/transport/uv/listener.cc:126 "rv < 0: invalid argument"
terminate called after throwing an instance of 'std::runtime_error'
  what():  In initFromLoop at ../third_party/tensorpipe/tensorpipe/transport/uv/listener.cc:126 "rv < 0: invalid argument"
```

**Comment by verdurin** [comment]
Test report by @verdurin
**FAILED**
Build succeeded for 1 out of 2 (1 easyconfigs in this PR)
easybuild.novalocal - Linux centos linux 7.8.2003, x86_64, Intel Xeon Processor (Skylake, IBRS), Python 2.7.5
See https://gist.github.com/5f838a139964afb0d53f51173e6da4a2 for a full test report.

**Comment by Flamefire** [comment]
Unfortunately no idea. Maybe check with PyTorch and/or search their issues if that is a known bug. Looks like libuv related. Exception is here: https://github.com/pytorch/tensorpipe/blob/779225daefc68b17a9729a37572f55fa1ff5b9fa/tensorpipe/transport/uv/listener.cc#L126

So maybe a firewall or something?

**Comment by verdurin** [comment]
Seeing the same errors on two unrelated systems on completely different networks - could be problem at the remote end, of course.

**Comment by Flamefire** [comment]
Just seen that this ain't my EC but the foss one. Were they actually requests for that? ML without GPUs is...slow.

@verdurin Does the fosscuda version work for you: https://github.com/easybuilders/easybuild-easyconfigs/pull/11041

**Comment by Flamefire** [comment]
Test report by @Flamefire
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
taurusml3 - Linux RHEL 7.6, POWER, 8335-GTX, Python 2.7.5
See https://gist.github.com/d03b18547a1b6522e8598f866b6bc8f6 for a full test report.

**Comment by Flamefire** [comment]
Test report by @Flamefire
**FAILED**
Build succeeded for 4 out of 5 (1 easyconfigs in this PR)
taurusa15 - Linux centos linux 7.7.1908, x86_64, Intel(R) Xeon(R) CPU E5-2603 v4 @ 1.70GHz, Python 2.7.5
See https://gist.github.com/d448153b8a3ae07d3d4d169dd536e34d for a full test report.

**Comment by terjekv** [comment]
Test report by @terjekv
**FAILED**
Build succeeded for 67 out of 72 (1 easyconfigs in this PR)
arm1 - Linux ubuntu 18.04, AArch64, UNKNOWN (aarch64), Python 3.6.9
See https://gist.github.com/b4f0f2a7cc9aac1860465502e666e06d for a full test report.

**Comment by Flamefire** [comment]
I further traced the issue on POWER: Seems a call to `std::exp` is causing a `FE_DIVBYZERO` or plain wrong results. Using some printf debugging I've seen that `exp(x[:] - max(x))` (Python pseudocode) transforms `in[-inf, -inf, ]` to `out[0, 1, ]` which is obviously wrong and even inconsistent.

This issue seems to reproducible go away when NOT using GLog. I don't see how that is related but that's the situation.

**Comment by Flamefire** [comment]
@boegel Can you please remove the glog and gflags dependencies?

**Comment by lexming** [comment]
Test report by @lexming
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
node128.hydra.os - Linux centos linux 7.7.1908, x86_64, Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz, Python 2.7.5
See https://gist.github.com/224e2bee0bad992ba217363980f5492e for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
node378.hydra.os - Linux centos linux 7.7.1908, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 2.7.5
See https://gist.github.com/3b618967573a2f8b5056087848c55008 for a full test report.

File: easybuild/easyconfigs/p/PyTorch/PyTorch-1.6.0-foss-2019b-Python-3.7.4.eb [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+    ('Python', '3.7.4'),
+    ('pybind11', '2.4.3', versionsuffix),
+    ('SciPy-bundle', '2019.10', versionsuffix),
+    ('PyYAML', '5.1.2'),
+    ('MPFR', '4.0.2'),
+    ('GMP', '6.1.2'),
+    ('numactl', '2.0.12'),
+    ('FFmpeg', '4.2.1'),
+    ('gflags', '2.2.2'),
+    ('glog', '0.4.0'),
```
Comment (lexming) [inline easybuild/easyconfigs/p/PyTorch/PyTorch-1.6.0-foss-2019b-Python-3.7.4.eb:277]: Please remove these dependencies as it was already done in https://github.com/easybuilders/easybuild-easyconfigs/pull/11323 for `fosscuda`
```suggestion
```

**Review by lexming — CHANGES_REQUESTED** [review]

**Comment by boegel** [comment]
@Flamefire Please retest on POWER?
@terjekv Please retest on Arm64?
@verdurin Please retest?

**Comment by boegel** [comment]
@boegelbot please test @ generoso

**Comment by terjekv** [comment]
Test report by @terjekv
**FAILED**
Build succeeded for 0 out of 6 (1 easyconfigs in this PR)
arm1 - Linux ubuntu 18.04, AArch64, UNKNOWN (aarch64), Python 3.6.9
See https://gist.github.com/a18425d1fb8f14a31ffb3ef010089d83 for a full test report.

**Comment by Flamefire** [comment]
Test report by @Flamefire
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
taurusml1 - Linux RHEL 7.6, POWER, 8335-GTX, Python 2.7.5
See https://gist.github.com/c1ea8e731756b546322839fae7481eba for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 7 out of 7 (1 easyconfigs in this PR)
node3125.skitty.os - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8
See https://gist.github.com/7b681741560321c8beae7d8846ea8abe for a full test report.

**Comment by Flamefire** [comment]
@boegel Currently messing a bit with OpenBLAS which might be the issue...

**Comment by Flamefire** [comment]
Test report by @Flamefire
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
taurusml7 - Linux RHEL 7.6, POWER, 8335-GTX, Python 2.7.5
See https://gist.github.com/e7004a4776c848793b1fd19e67e72469 for a full test report.

**Comment by Flamefire** [comment]
Test report by @Flamefire
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
taurusml4 - Linux RHEL 7.6, POWER, 8335-GTX, Python 2.7.5
See https://gist.github.com/974245e2ddd9b64ede1fc4f58cf071bf for a full test report.

**Comment by Flamefire** [comment]
Ok, no idea if I messed up in my env or a flaky test. Anyway, :+1: from me

**Comment by lexming** [comment]
Test report by @lexming
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in this PR)
node368.hydra.os - Linux centos linux 7.7.1908, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 2.7.5
See https://gist.github.com/8ab3c2fbf0bad362460dda8d6081790a for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
node101.hydra.os - Linux centos linux 7.7.1908, x86_64, Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz, Python 2.7.5
See https://gist.github.com/89f5b0e6258d9014280297a15c9a1c91 for a full test report.

**Review by lexming — APPROVED** [review]
LGTM

**Comment by lexming** [comment]
Since the only failed tests are from @terjekv on ARM, I think that we can merge this. Fixes for ARM can come in a subsequent PR.

Going in, thanks @boegel !

**Comment by terjekv** [comment]
Test report by @terjekv
**FAILED**
Build succeeded for 62 out of 75 (1 easyconfigs in this PR)
arm2 - Linux ubuntu 18.04, AArch64, UNKNOWN, Python 3.6.9
See https://gist.github.com/d225e145abd93914903dbfbdebf4249f for a full test report.

**Comment by terjekv** [comment]
@boegel?
```
== 2020-10-06 18:13:08,018 modules.py:614 INFO Module Pillow/6.2.1-GCCcore-8.3.0 not found in list of available modules, checking via 'module show'...                        │···
== 2020-10-06 18:13:08,097 modules.py:627 INFO Module Pillow/6.2.1-GCCcore-8.3.0 not found via module avail/show, checking whether it is a wrapper                            │···
== 2020-10-06 18:13:08,100 modules.py:634 INFO Result for existence check of Pillow/6.2.1-GCCcore-8.3.0 module: False                                                         │···
== 2020-10-06 18:13:08,105 build_log.py:169 ERROR EasyBuild crashed with an error (at easybuild/software/EasyBuild/4.3.0/lib/python3.6/site-packages/easybuild/base/exceptions│···
.py:124 in __init__): Missing modules for dependencies (use --robot?): hypothesis/4.44.2-GCCcore-8.3.0-Python-3.7.4, Ninja/1.9.0-GCCcore-8.3.0, protobuf-python/3.10.0-foss-20│···
19b-Python-3.7.4, Python/3.7.4-GCCcore-8.3.0, pybind11/2.4.3-GCCcore-8.3.0-Python-3.7.4, SciPy-bundle/2019.10-foss-2019b-Python-3.7.4, PyYAML/5.1.2-GCCcore-8.3.0, Pillow/6.2.│···
1-GCCcore-8.3.0 (at easybuild/software/EasyBuild/4.3.0/lib/python3.6/site-packages/easybuild/tools/toolchain/toolchain.py:555 in _check_dependencies)                         │···
```

**Comment by Flamefire** [comment]
@terjekv OpenBLAS build failed: 
```
gcc -O2 -ftree-vectorize -mcpu=native -fno-math-errno -O2 -DMAX_STACK_ALLOC=2048 -fopenmp -Wall -DF_INTERFACE_GFORT -fPIC -DSMP_SERVER -DUSE_OPENMP -DNO_WARMUP -DMAX_CPU_NUMBER=96 -DMAX_PARALLEL_NUMBER=1 -DVERSION=\"0.3.7\" -march=armv8.2-a -mtune=tsv110 -DASMNAME=ismax -DASMFNAME=ismax_ -DNAME=ismax_ -DCNAME=ismax -DCHAR_NAME=\"ismax_\" -DCHAR_CNAME=\"ismax\" -DNO_AFFINITY -I.. -I. -UDOUBLE  -UCOMPLEX -c -UUSE_ABS -UUSE_MIN imax.c -o ismax.o
cc1: error: unknown value tsv110 for -mtune
cc1: note: valid arguments are: cortex-a35 cortex-a53 cortex-a57 cortex-a72 cortex-a73 thunderx thunderxt88p1 thunderxt88 thunderxt81 thunderxt83 xgene1 falkor qdf24xx exynos-m1 thunderx2t99p1 vulcan thunderx2t99 cortex-a55 cortex-a75 saphira cortex-a57.cortex-a53 cortex-a72.cortex-a53 cortex-a73.cortex-a35 cortex-a73.cortex-a53 cortex-a75.cortex-a55 generic
```

I guess that gcc is to old for your tune setting? Seems that feature was added in GCC 9.1: https://github.com/gcc-mirror/gcc/commit/910f72e78fe76993b8a9bcc9e385a788446f1f10

Seems you could use `TARGET=CORTEXA57` for build/install/testopts

**Comment by terjekv** [comment]
Right. We should probably detect that? :-/

**Comment by Flamefire** [comment]
Not sure if we can do that automatically... Probably a patch for that EC would be better which makes it use the A57 values for tsv110 instead

**Comment by terjekv** [comment]
I'm not sure why OpenBLAS only fails with this build, but yeah, it works everywhere else. Have it break on arm and have someone (moi?) fix it eventually.

**Comment by boegel** [comment]
```
cc1: error: unknown value tsv110 for -mtune
```

**Comment by Flamefire** [comment]
> I'm not sure why OpenBLAS only fails with this build, but yeah, it works everywhere else. Have it break on arm and have someone (moi?) fix it eventually.

The arch options used by OpenBLAS are to new for GCC 8. It needs a patch as mentioned in https://github.com/easybuilders/easybuild-easyconfigs/pull/11251#issuecomment-704909724 or https://github.com/easybuilders/easybuild-easyconfigs/pull/11251#issuecomment-704948375

**Comment by Flamefire** [comment]
Added a PR to resolve this

**Pull request merged**